### ISSUE TITLE
Getting useful errors when metrics fail - Header size exceeded max allowed size

### ISSUE THREAD

**Issue by pulewicz** [GoogleCloudPlatform/opentelemetry-operations-java#149]
Hi,

We've added google metrics exporter to our codebase. It works locally, however when we deploy it on GKE, we get the following error (full stacktrace at the end of the post):

```
io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2Exception: Header size exceeded max allowed size (10240)
```

I think what's happening is we get some HTTP error response, but we can't decode its because of the header size limit. Is it possible to increase the limit or get more information about this error some other way?

Full stacktrace:

```
[otel.javaagent 2022-04-28 18:05:34:341 +0000] [PeriodicMetricReader-1] WARN io.opentelemetry.sdk.metrics.export.PeriodicMetricReader - Exporter threw an Exception
com.google.api.gax.rpc.InternalException: io.grpc.StatusRuntimeException: INTERNAL: http2 exception
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:67)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60)
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:68)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1074)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1213)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:983)
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:771)
	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:563)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:533)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:553)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:68)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:739)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:718)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
	Suppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
		at com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)
		at com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112)
		at com.google.cloud.monitoring.v3.MetricServiceClient.createTimeSeries(MetricServiceClient.java:1526)
		at com.google.cloud.monitoring.v3.MetricServiceClient.createTimeSeries(MetricServiceClient.java:1469)
		at com.google.cloud.opentelemetry.metric.CloudMetricClientImpl.createTimeSeries(CloudMetricClientImpl.java:39)
		at com.google.cloud.opentelemetry.metric.MetricExporter.createTimeSeriesBatch(MetricExporter.java:210)
		at com.google.cloud.opentelemetry.metric.MetricExporter.export(MetricExporter.java:195)
		at io.opentelemetry.sdk.metrics.export.PeriodicMetricReader$Scheduled.doRun(PeriodicMetricReader.java:125)
		at io.opentelemetry.sdk.metrics.export.PeriodicMetricReader$Scheduled.run(PeriodicMetricReader.java:117)
		at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
		at java.base/java.util.concurrent.FutureTask.runAndReset(Unknown Source)
		at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
		... 3 more
Caused by: io.grpc.StatusRuntimeException: INTERNAL: http2 exception
	at io.grpc.Status.asRuntimeException(Status.java:535)
	... 10 more
Caused by: io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2Exception: Header size exceeded max allowed size (10240)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2Exception.connectionError(Http2Exception.java:103)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2CodecUtil.headerListSizeExceeded(Http2CodecUtil.java:245)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader$HeadersBlockBuilder.headerSizeExceeded(DefaultHttp2FrameReader.java:694)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader$HeadersBlockBuilder.addFragment(DefaultHttp2FrameReader.java:710)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader$2.processFragment(DefaultHttp2FrameReader.java:481)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader.readHeadersFrame(DefaultHttp2FrameReader.java:491)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader.processPayloadState(DefaultHttp2FrameReader.java:254)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2FrameReader.readFrame(DefaultHttp2FrameReader.java:160)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2InboundFrameLogger.readFrame(Http2InboundFrameLogger.java:41)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.decodeFrame(DefaultHttp2ConnectionDecoder.java:174)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2ConnectionHandler$FrameDecoder.decode(Http2ConnectionHandler.java:378)
	at io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2ConnectionHandler.decode(Http2ConnectionHandler.java:438)
	at io.grpc.netty.shaded.io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:501)
	at io.grpc.netty.shaded.io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:440)
	at io.grpc.netty.shaded.io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.grpc.netty.shaded.io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1526)
	at io.grpc.netty.shaded.io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1275)
	at io.grpc.netty.shaded.io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1322)
	at io.grpc.netty.shaded.io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:501)
	at io.grpc.netty.shaded.io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:440)
	at io.grpc.netty.shaded.io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.grpc.netty.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.grpc.netty.shaded.io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.grpc.netty.shaded.io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:795)
	at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:475)
	at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378)
	at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.grpc.netty.shaded.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.grpc.netty.shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	... 1 more
```

Regards,
Paweł Ulewicz.

**Comment by elefeint** [comment]
GoogleCloudPlatform/spring-cloud-gcp#1223 may be of interest here. Not sure why duplicate metrics names results in a generic `Header size exceeded max allowed size` response.

**Comment by AQS-DTheuke** [comment]
The issue there (and maybe here) is that the error message is written to the headers (formatted for better readability):

````txt
com.google.api.gax.rpc.InvalidArgumentException: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: 
One or more TimeSeries could not be written: 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[45]: custom.googleapis.com/jdbc/connections/max{namedataSource}; 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[47]: custom.googleapis.com/process/start/time{}; 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[74]: custom.googleapis.com/hikaricp/connections/acquire/max{poolHikariPool-1}; 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[2,55,56,68,72,79,117,120]: custom.googleapis.com/jvm/memory/committed{idTenured Gen,areaheap}; 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[70]: custom.googleapis.com/jvm/threads/peak{}; 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[13]: custom.googleapis.com/tomcat/sessions/active/current{}; 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[67]: custom.googleapis.com/jvm/classes/unloaded{}; 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[41]: custom.googleapis.com/jvm/memory/usage/after/gc{areaheap,poollong-lived}; 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[49]: custom.googleapis.com/jvm/threads/daemon{}; 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[17]: custom.googleapis.com/system/cpu/usage{}; 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[8]: custom.googleapis.com/hikaricp/connections/usage{poolHikariPool-1}; 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[76]: custom.googleapis.com/hikaricp/connections/min{poolHikariPool-1}; 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[69]: custom.googleapis.com/jvm/classes/loaded{}; 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[58]: custom.googleapis.com/process/files/open{}; 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[59]: custom.googleapis.com/jvm/gc/live/data/size{}; 
One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[85]: custom.googleapis.com/jvm/gc/memory/allocated{}
````

This is 3KB for 16 failed metrics. If the application has slightly more time series that cause a conflict, then the error message seems to exceed the header limit.
(I have read somewhere that the error message is duplicated for some reason.)

You may also want to read this comment:
https://github.com/micrometer-metrics/micrometer/issues/1268#issuecomment-1210153226

TLDR: This is probably an error/bug/incompatibility with the (error-)response sent by the google servers/services.

**Comment by aabmass** [comment]
Thanks for looking into issue.

> TLDR: Configure the underlying grpc client to accept larger header sizes (`maxInboundMetadataSize`) (effectively turning off client side DOS protections?).

I don't think we would want to do this by default in the exporter. We could possibly expose this as an option that we pass through so folks can opt into a fix.

Maybe someone should open a bug against https://github.com/googleapis/java-monitoring, which we are using in this exporter and is where the issue is originating from.

---

Re the error you're receiving

> ```
> One or more points were written more frequently than the maximum sampling period configured for the metric.: global{} timeSeries[45]: custom.googleapis.com/jdbc/connections/max{namedataSource};
> ```

this is a common issue, and there's a few different potential causes. If you're still being hit by this, would you mind opening a separate issue in this repo to discuss?

**Comment by damemi** [comment]
Update: @psx95 is still planning to work on this, but we haven't had the capacity yet

**Comment by Mowee** [comment]
@damemi @psx95 Is there an update to this issue? Not sure if this this the right place where this needs to be fixed but this keeps appearing in our logs.

**Comment by psx95** [comment]
@Mowee, I recently started looking into this issue again and based on the investigation already done on this issue and some additional testing, we are looking into the feasibility of exposing an option to configure the [maxInboundMetadataSize](https://grpc.github.io/grpc-java/javadoc/io/grpc/ManagedChannelBuilder.html#maxInboundMetadataSize(int)) for the client used to make export calls.
